magicctapipe/scripts/lst1_magic/semi_automatic_scripts/database_production/update_MAGIC_database.py

```diff
@@ -0,0 +1,351 @@
+        DF : pandas.DataFrame
+            The previous MAGIC database which we want to update.
+        tel_id : int
+            The telescope ID, which must be either 1 or 2.
+
+    Returns
+    -------
+    pandas.DataFrame
+        A DataFrame with updated MAGIC database.
+    """
+
+    database["DATE"] = database["DATE"].astype(str)
+    DF["DATE"] = DF["DATE"].astype(str)
+    columns_to_compare = ["DATE", "source", "Run ID"]
+    merged_df = pd.merge(
+        database,
+        DF[columns_to_compare],
+        on=columns_to_compare,
+        how="left",
+        indicator=True,
+    )
+    non_matching_rows = merged_df[merged_df["_merge"] == "left_only"].drop(
+        columns=["_merge"]
+    )
+
+    if non_matching_rows.empty:
+        raise Exception("There is no un-updated data for a given time interval. ")
+    else:
+
+        non_matching_rows_reset = non_matching_rows.reset_index(drop=True)
+        new_rows = []
+
+        for index, row in non_matching_rows_reset.iterrows():
+            date = row["DATE"]
+            source = row["source"]
+            run_id = row["Run ID"]
+            run_id = str(run_id)
+
+            date_obj = datetime.strptime(date, "%Y%m%d")
+            date_obj += timedelta(days=1)
+            new_date = datetime.strftime(date_obj, "%Y%m%d")
+            YYYY = new_date[:4]
+            MM = new_date[4:6]
+            DD = new_date[6:8]
+            Y = "_Y_"
+
+            path = f"/fefs/onsite/common/MAGIC/data/M{tel_id}/event/Calibrated/{YYYY}/{MM}/{DD}"
+
+            if os.path.exists(path):
+                files = os.listdir(path)
+                count_with_run_id = 0
+                # Counter for files that include the run_id.
+                for filename in files:
+                    if Y in filename:
+                        if new_date in filename:
+                            if source in filename:
+                                if run_id in filename:
+                                    count_with_run_id += 1
+                if count_with_run_id != 0:
+                    new_rows.append(
+                        {
+                            "DATE": date,
+                            "source": source,
+                            "Run ID": run_id,
+                            "number of subruns": count_with_run_id,
+                        }
+                    )
+
+        new_rows = pd.DataFrame(new_rows)
+        new_rows["DATE"] = pd.to_datetime(new_rows["DATE"])
+        combined_df = pd.concat([DF, new_rows], ignore_index=True)
+        combined_df["DATE"] = pd.to_datetime(combined_df["DATE"], errors="coerce")
+        combined_df = combined_df.sort_values("DATE")
+
+        combined_df["DATE"] = combined_df["DATE"].dt.strftime("%Y%m%d")
+        combined_df["Run ID"] = combined_df["Run ID"].astype(int)
+        combined_df.reset_index(drop=True, inplace=True)
+
+        for column in combined_df.columns[4:]:
```

Comment:
this also loops over the "old" lines in the table, and some of those might not be processed yet, did you check if there are no problems caused by this?

Reply:
Yes, this loop iterates over each cell/line in a column. The aim of this part of the code is to avoid discrepancies in data types before saving a file as HDF5. I assumed that if a file has not been processed yet, the default value in a cell would be NaN. Then, lines with NaN are skipped, and we look for a cell with a value in the column to check its type. This type is then used to define the rest of the column with a new default value. I hope that the default value used is neutral enough. If the entire column contains NaN (indicating that no files have been processed yet), we skip the column, avoiding any issues with data types. However, after your question, I realized that a cell could contain an empty string or only whitespace. I will take this into account when updating the code.

Reply:
I understand, but without this step, I am unable to save my tables to an HDF5 file due to mixed data types. For example, I encountered the following error: 'An error occurred: Cannot serialize the column [DL1_rc_all] because its data contents are not [string] but [mixed] object dtype.

Reply:
After resolving this issue, we should be able to merge the branches, as the current version has no linting, pyflakes, or docs warnings.